src/tree/zk.rs

```diff
@@ -15,65 +15,227 @@ pub fn merkle_opening<const DEPTH: usize>(
     branch: &PoseidonBranch<DEPTH>,
     leaf: Witness,
 ) -> Witness {
-    let zero = TurboComposer::constant_zero();
-
-    let mut base = true;
-    let mut root = zero;
-
     // Generate a permutation container
-    let mut perm = [zero; dusk_hades::WIDTH];
+    let mut container = [TurboComposer::constant_zero(); dusk_hades::WIDTH];
+
+    // Recalculate the root for the given branch
+    (0..DEPTH).fold(leaf, |root, depth| {
+        let level = branch.as_ref()[depth];
 
-    // For every level, replace the level offset with needle,
-    // permute the level and set the needle to the next level
-    // to the poseidon result of the permutation
-    branch.as_ref().iter().for_each(|level| {
         // Create the bits representation of the offset as witness
+        // and make sure that offset points to a hash in the level
         let offset_flag = level.offset_flag();
-
-        let mut sum = zero;
-        let mut offset_bits = [zero; dusk_hades::WIDTH - 1];
+        let mut sum = TurboComposer::constant_zero();
+        let mut offset_bits =
+            [TurboComposer::constant_zero(); dusk_hades::WIDTH - 1];
         offset_bits.iter_mut().fold(1, |mask, bit| {
-            let b = BlsScalar::from((offset_flag & mask).min(1));
-            *bit = composer.append_witness(b);
+            let bit_bls = BlsScalar::from((offset_flag & mask).min(1));
+            *bit = composer.append_witness(bit_bls);
 
             let constraint = Constraint::new().left(1).a(sum).right(1).b(*bit);
             sum = composer.gate_add(constraint);
 
             mask << 1
         });
-
         composer.assert_equal_constant(sum, BlsScalar::one(), None);
 
-        let needle = composer.append_witness(**level);
-        if base {
-            composer.assert_equal(leaf, needle);
-            base = false;
+        // Check that the root of the previous level is the same value as what
+        // is stored in the level at the offset
+        for i in 0..dusk_hades::WIDTH {
+            // Load child hashes of the current level into the permutation
+            // container
+            container[i] = composer.append_witness(level.as_ref()[i]);
+            if i > 0 {
+                let bit = offset_bits[i - 1];
+
+                // `expected` and `calculated` will be zero everywhere except
+                // at the level offset
+                let constraint =
+                    Constraint::new().mult(1).a(bit).b(container[i]);
+                let expected = composer.gate_mul(constraint);
+                let constraint = Constraint::new().mult(1).a(bit).b(root);
+                let calculated = composer.gate_mul(constraint);
+
+                // Make sure that the hash at the offset of the current level
+                // matches the hash calculated in the previous level
+                composer.assert_equal(expected, calculated);
```

Comment:
The `expected` here is a new witness, which can be anything. We need a way to assert that this new witness corresponds to the previous root